File: LedgeriOS/LedgerTargetCore/ClientSummaryPhysicalReport.swift
import Foundation

public enum ClientSummaryPhysicalReportUpdate: Equatable, Sendable {
    case incomplete
    case ready(ClientSummaryPhysicalReportSnapshot)
}

public protocol ClientSummaryPhysicalReportWatching: Sendable {
    func watchClientSummaryPhysicalReport(accountId: AccountID, projectId: ProjectID)
        -> AsyncThrowingStream<ClientSummaryPhysicalReportUpdate, Error>
}

public protocol ClientSummaryPhysicalReportReading: Sendable {
    func readDownloadedClientSummaryPhysicalReport(accountId: AccountID, projectId: ProjectID,
        asOf: ProtectedArtifactEpochMilliseconds) async throws -> ClientSummaryPhysicalReportSnapshot
}

public enum ClientSummaryPhysicalReportClient: Encodable, Equatable, Sendable {
    case known(clientId: ClientID, name: String, revision: UInt64)
    case unavailable(clientId: ClientID)

    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .known(let id, let name, let revision):
            try values.encode("known", forKey: .kind)
            try values.encode(id, forKey: .clientId)
            try values.encode(name, forKey: .name)
            try values.encode(String(revision), forKey: .revision)
        case .unavailable(let id):
            try values.encode("unavailable", forKey: .kind)
            try values.encode(id, forKey: .clientId)
        }
    }
    private enum CodingKeys: String, CodingKey { case kind, clientId, name, revision }
}

/// Missing association evidence is not an uncategorized Item. The reader must
/// supply the actual physical Item category identity, never an Invoice fallback.
public enum ClientSummaryPhysicalReportCategory: Encodable, Equatable, Sendable {
    case known(categoryId: BudgetCategoryID, name: String)
    case unavailable
}

public struct ClientSummaryPhysicalReportItem: Encodable, Equatable, Sendable {
    public let accountId: AccountID
    public let projectId: ProjectID
    public let itemId: ItemID
    public let placementId: EntityID
    public let spaceId: SpaceID?
    public let name: String
    public let sku: String?
    public let category: ClientSummaryPhysicalReportCategory
    public let itemRevision: UInt64
    public let accounting: ProjectItemAccountingRow?

    public init(accountId: AccountID, projectId: ProjectID, itemId: ItemID,
                placementId: EntityID, spaceId: SpaceID?, name: String, sku: String?,
                category: ClientSummaryPhysicalReportCategory, itemRevision: UInt64,
                accounting: ProjectItemAccountingRow? = nil) {
        self.accountId = accountId; self.projectId = projectId; self.itemId = itemId
        self.placementId = placementId; self.spaceId = spaceId; self.name = name
        self.sku = sku; self.category = category; self.itemRevision = itemRevision
        self.accounting = accounting
    }

    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(accountId, forKey: .accountId)
        try values.encode(projectId, forKey: .projectId)
        try values.encode(itemId, forKey: .itemId)
        try values.encode(placementId, forKey: .placementId)
        try values.encode(spaceId, forKey: .spaceId)
        try values.encode(name, forKey: .name)
        try values.encode(sku, forKey: .sku)
        try values.encode(category, forKey: .category)
        try values.encode(String(itemRevision), forKey: .itemRevision)
        try values.encode(accounting, forKey: .accounting)
    }
    private enum CodingKeys: String, CodingKey {
        case accountId, projectId, itemId, placementId, spaceId, name, sku, category, itemRevision, accounting
    }
}

public enum ClientSummaryPhysicalReportFailure: Error, Equatable, Sendable {
    case incompleteReadiness, scopeMismatch, invalidRevision
    case duplicateItem, duplicatePlacement, duplicateSpace, missingSpace, conflictingCategory
}

/// Physical detail only. Financial totals, budgets and receipt links are separate
/// governed capabilities and are deliberately absent from this payload.
public struct ClientSummaryPhysicalReportSnapshot: Encodable, Equatable, Sendable {
    public let reference: ProtectedArtifactSnapshotReference
    public let project: PropertyManagementReportProject
    public let client: ClientSummaryPhysicalReportClient
    public let spaces: [PropertyManagementReportSpace]
    public let items: [ClientSummaryPhysicalReportItem]
    public let provenance: PropertyManagementReportProvenance
    public let sourceSetHash: ProtectedArtifactSHA256
    public let accountingEvidenceHash: ProtectedArtifactSHA256
    public let reportKind = "client_summary_physical"

    /// Incomplete detail may be previewed, but cannot become a completed export.
    public var isComplete: Bool {
        guard case .known = client else { return false }
        return items.allSatisfy {
            guard $0.accounting?.resolution == .accountedFor else { return false }
            if case .known = $0.category { return true }; return false
        }
    }

    public static func build(project: PropertyManagementReportProject,
                             client: ClientSummaryPhysicalReportClient,
                             spaces: [PropertyManagementReportSpace],
                             items: [ClientSummaryPhysicalReportItem],
                             provenance: PropertyManagementReportProvenance) throws -> Self {
        guard provenance.readiness == .ready else { throw ClientSummaryPhysicalReportFailure.incompleteReadiness }
        guard project.accountId == provenance.accountId, project.projectId == provenance.projectId else {
            throw ClientSummaryPhysicalReportFailure.scopeMismatch
        }
        guard project.revision > 0 else { throw ClientSummaryPhysicalReportFailure.invalidRevision }
        if case .known(_, let name, let revision) = client {
            guard revision > 0 else { throw ClientSummaryPhysicalReportFailure.invalidRevision }
            _ = try ClientDisplayName(validating: name)
        }
        var spaceIDs = Set<SpaceID>()
        for space in spaces {
            guard space.accountId == project.accountId, space.projectId == project.projectId else {
                throw ClientSummaryPhysicalReportFailure.scopeMismatch
            }
            guard space.revision > 0 else { throw ClientSummaryPhysicalReportFailure.invalidRevision }
            guard spaceIDs.insert(space.spaceId).inserted else { throw ClientSummaryPhysicalReportFailure.duplicateSpace }
        }
        var itemIDs = Set<ItemID>(), placementIDs = Set<EntityID>()
        var categoryNames: [BudgetCategoryID: String] = [:]
        let clientId: ClientID
        switch client {
        case .known(let id, _, _), .unavailable(let id): clientId = id
        }
        for item in items {
            guard item.accountId == project.accountId, item.projectId == project.projectId else {
                throw ClientSummaryPhysicalReportFailure.scopeMismatch
            }
            guard item.itemRevision > 0 else { throw ClientSummaryPhysicalReportFailure.invalidRevision }
            if let evidence = item.accounting?.evidence {
                guard evidence.accountId == item.accountId,
                      evidence.projectId == item.projectId,
                      evidence.clientId == clientId,
                      evidence.itemId == item.itemId,
                      evidence.spaceId == item.spaceId else {
                    throw ClientSummaryPhysicalReportFailure.scopeMismatch
                }
            }
            guard itemIDs.insert(item.itemId).inserted else { throw ClientSummaryPhysicalReportFailure.duplicateItem }
            guard placementIDs.insert(item.placementId).inserted else { throw ClientSummaryPhysicalReportFailure.duplicatePlacement }
            if let space = item.spaceId, !spaceIDs.contains(space) { throw ClientSummaryPhysicalReportFailure.missingSpace }
            if case .known(let id, let name) = item.category {
                _ = try BudgetCategoryName(validating: name)
                if let previous = categoryNames[id], !previous.utf8.elementsEqual(name.utf8) {
                    throw ClientSummaryPhysicalReportFailure.conflictingCategory
                }
                categoryNames[id] = name
            }
        }
        func ordered(_ a: String, _ aid: String, _ b: String, _ bid: String) -> Bool {
            a.utf8.elementsEqual(b.utf8) ? aid.utf8.lexicographicallyPrecedes(bid.utf8) : a.utf8.lexicographicallyPrecedes(b.utf8)
        }
        let spaces = spaces.sorted { ordered($0.name, $0.spaceId.rawValue, $1.name, $1.spaceId.rawValue) }
        let items = items.sorted { ordered($0.name, $0.itemId.rawValue, $1.name, $1.itemId.rawValue) }
        // Commit the eligibility evidence, including excluded identities, without
        // putting Unaccounted For physical detail in the report payload.
        let accountingEvidenceHash = try ProtectedArtifactSHA256.make(bytes: encode(items.sorted {
            $0.itemId.rawValue.utf8.lexicographicallyPrecedes($1.itemId.rawValue.utf8)
        }.map {
            AccountingSource(itemId: $0.itemId, accounting: $0.accounting)
        }))
        let reportItems = items.filter { $0.accounting?.resolution != .unaccountedFor }
        let source = Source(project: project, client: client, spaces: spaces, items: reportItems,
                            accountingEvidenceHash: accountingEvidenceHash)
        let sourceHash = try ProtectedArtifactSHA256.make(bytes: encode(source))
        let content = Content(reportKind: "client_summary_physical", source: source, provenance: provenance, sourceSetHash: sourceHash)
        let hash = try ProtectedArtifactSHA256.make(bytes: encode(content))
        let reference = try ProtectedArtifactSnapshotReference(snapshotID: .init(validating: String(hash.rawValue.prefix(32))),
            snapshotHash: hash, visibilityScopeID: provenance.visibilityScopeID,
            profileVersion: .init(validating: "client-summary-physical-v1"), authorityVersion: provenance.authorityVersion)
        return .init(reference: reference, project: project, client: client, spaces: spaces,
                     items: reportItems, provenance: provenance, sourceSetHash: sourceHash,
                     accountingEvidenceHash: accountingEvidenceHash)
    }

    public func canonicalData() throws -> Data { try Self.encode(self) }
    public func canonicalContentData() throws -> Data {
        try Self.encode(Content(reportKind: reportKind,
            source: Source(project: project, client: client, spaces: spaces, items: items,
                           accountingEvidenceHash: accountingEvidenceHash),
            provenance: provenance, sourceSetHash: sourceSetHash))
    }
    private static func encode<T: Encodable>(_ value: T) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(value)
    }
    private struct Source: Encodable {
        let project: PropertyManagementReportProject
        let client: ClientSummaryPhysicalReportClient
        let spaces: [PropertyManagementReportSpace]
        let items: [ClientSummaryPhysicalReportItem]
        let accountingEvidenceHash: ProtectedArtifactSHA256
    }
    private struct AccountingSource: Encodable {
        let itemId: ItemID
        let accounting: ProjectItemAccountingRow?
    }
    private struct Content: Encodable {
        let reportKind: String
        let source: Source
        let provenance: PropertyManagementReportProvenance
        let sourceSetHash: ProtectedArtifactSHA256
    }
}

File: LedgerTargetMCP/src/clientSummaryPhysicalReport.ts
import { TargetMCPFailure } from "./contractSupport.js";
import { propertyReportSupport as v, type PropertyManagementReportProject,
  type PropertyManagementReportSpace, type AuthoritativePropertyManagementReportProvenance,
  type ProjectItemAccountingRow } from "./propertyManagementReport.js";

export type ClientSummaryPhysicalReportClient =
  | Readonly<{ kind: "known"; clientId: string; name: string; revision: string }>
  | Readonly<{ kind: "unavailable"; clientId: string }>;
// Swift's synthesized enum encoding, unlike the Client's explicit encoding.
export type ClientSummaryPhysicalReportCategory =
  | Readonly<{ known: Readonly<{ categoryId: string; name: string }> }>
  | Readonly<{ unavailable: Readonly<Record<string, never>> }>;
export type ClientSummaryPhysicalReportItem = Readonly<{
  accountId: string; projectId: string; itemId: string; placementId: string;
  spaceId: string | null; name: string; sku: string | null;
  category: ClientSummaryPhysicalReportCategory; itemRevision: string;
  accounting: ProjectItemAccountingRow | null;
}>;
export type ClientSummaryPhysicalReportInput = Readonly<{
  project: PropertyManagementReportProject; client: ClientSummaryPhysicalReportClient;
  spaces: readonly PropertyManagementReportSpace[]; items: readonly ClientSummaryPhysicalReportItem[];
  provenance: AuthoritativePropertyManagementReportProvenance;
}>;
export type ClientSummaryPhysicalReportSnapshot = ClientSummaryPhysicalReportInput & Readonly<{
  reportKind: "client_summary_physical"; sourceSetHash: string; accountingEvidenceHash: string;
  reference: Readonly<{ snapshotID: string; snapshotHash: string; visibilityScopeID: string;
    profileVersion: "client-summary-physical-v1"; authorityVersion: string }>;
}>;
function fail(code: string): never { throw new TargetMCPFailure(`client_summary_physical_${code}`); }
// Foundation whitespacesAndNewlines includes NEL and zero-width space; JS trim
// differs on both and on BOM. Preserve original spelling in the report payload.
const edgeWhitespace = /^[\p{White_Space}\u200b]+|[\p{White_Space}\u200b]+$/gu;
function name(value: unknown, category = false): string {
  const original = v.string(value), trimmed = original.replace(edgeWhitespace, "");
  if (!trimmed) fail("invalid_name");
  if (category && (/[\p{Cc}\p{Cf}]/u.test(trimmed)
    || [...new Intl.Segmenter("en", { granularity: "grapheme" }).segment(trimmed)].length > 100)) fail("invalid_name");
  return original;
}
function category(value: unknown): ClientSummaryPhysicalReportCategory {
  const raw = v.record(value);
  if (Object.keys(raw).length !== 1) fail("invalid_category");
  if (Object.hasOwn(raw, "known")) {
    const known = v.record(raw.known);
    return { known: { categoryId: v.id(known.categoryId), name: name(known.name, true) } };
  }
  if (Object.hasOwn(raw, "unavailable")) {
    v.record(raw.unavailable);
    return { unavailable: {} };
  }
  return fail("invalid_category");
}

/** Pure physical projection. The caller must authenticate and obtain a coherent
 * authorized read. Unknown evidence is retained for preview, never promoted. */
export function buildClientSummaryPhysicalReportSnapshot(input: ClientSummaryPhysicalReportInput): ClientSummaryPhysicalReportSnapshot {
  v.record(input); v.record(input.project); v.record(input.provenance);
  const project: PropertyManagementReportProject = { accountId: v.id(input.project.accountId),
    projectId: v.id(input.project.projectId), name: v.string(input.project.name),
    address: v.nullableText(input.project.address), revision: v.revision(input.project.revision) };
  const raw = input.provenance;
  v.scope(raw.accountId, raw.projectId, project);
  if (raw.readiness !== "ready") fail("incomplete_readiness");
  if (raw.source?.kind !== "authoritative" || Object.keys(raw.source).some(key => key !== "kind")
    || "lastSyncedAt" in raw || "localDataVersion" in raw) fail("invalid_source");
  if (typeof raw.visibilityScopeID !== "string" || !/^[a-f0-9]{64}$/.test(raw.visibilityScopeID)) fail("invalid_visibility_scope");
  if (typeof raw.authorityVersion !== "string" || !/^[a-z0-9_.-]{1,64}$/.test(raw.authorityVersion)) fail("invalid_authority_version");
  if (!Number.isSafeInteger(raw.asOf) || raw.asOf <= 0) fail("invalid_timestamp");
  const provenance: AuthoritativePropertyManagementReportProvenance = {
    accountId: project.accountId, projectId: project.projectId, principalId: v.id(raw.principalId),
    visibilityScopeID: raw.visibilityScopeID, source: { kind: "authoritative" },
    authorityVersion: raw.authorityVersion, asOf: raw.asOf, readiness: "ready" };
  const clientRaw = v.record(input.client), clientId = v.id(clientRaw.clientId);
  let client: ClientSummaryPhysicalReportClient;
  if (clientRaw.kind === "known") client = { kind: "known", clientId,
    name: name(clientRaw.name), revision: v.revision(clientRaw.revision) };
  else if (clientRaw.kind === "unavailable") client = { kind: "unavailable", clientId };
  else return fail("invalid_client");
  if (!Array.isArray(input.spaces) || !Array.isArray(input.items)) fail("invalid_rows");
  const spaceIDs = new Set<string>(), itemIDs = new Set<string>(), placementIDs = new Set<string>();
  const spaces = input.spaces.map(row => {
    v.record(row); v.scope(row.accountId, row.projectId, project);
    const spaceId = v.id(row.spaceId); v.unique(spaceIDs, spaceId, "duplicate_space");
    return { accountId: project.accountId, projectId: project.projectId, spaceId,
      name: v.string(row.name), revision: v.revision(row.revision) };
  }).sort((a, b) => v.ordered(a.name, a.spaceId, b.name, b.spaceId));
  const names = new Map<string, string>();
  const items: ClientSummaryPhysicalReportItem[] = input.items.map(row => {
    v.record(row); v.scope(row.accountId, row.projectId, project);
    const itemId = v.id(row.itemId), placementId = v.id(row.placementId);
    v.unique(itemIDs, itemId, "duplicate_item"); v.unique(placementIDs, placementId, "duplicate_placement");
    const spaceId = row.spaceId === null ? null : v.id(row.spaceId);
    if (spaceId !== null && !spaceIDs.has(spaceId)) fail("missing_space");
    const item = { accountId: project.accountId, projectId: project.projectId, itemId, placementId,
      spaceId, name: v.string(row.name), sku: v.nullableText(row.sku), itemRevision: v.revision(row.itemRevision) };
    const accounting = row.accounting == null ? null : v.accounting(row.accounting, item, true);
    if (accounting && accounting.evidence.clientId !== clientId) fail("scope_mismatch");
    const attribution = category(row.category);
    if ("known" in attribution) {
      const { categoryId, name } = attribution.known, previous = names.get(categoryId);
      if (previous !== undefined && previous !== name) fail("conflicting_category");
      names.set(categoryId, name);
    }
    return { ...item, category: attribution, accounting };
  }).sort((a, b) => v.ordered(a.name, a.itemId, b.name, b.itemId));
  // AccountingSource uses synthesized Optional encoding: nil is omitted here,
  // while ClientSummaryPhysicalReportItem explicitly encodes accounting:null.
  const accountingEvidenceHash = v.hash(v.canonical([...items]
    .sort((a, b) => v.ordered("", a.itemId, "", b.itemId))
    .map(item => ({ itemId: item.itemId, ...(item.accounting === null ? {} : { accounting: item.accounting }) }))));
  const reportItems = items.filter(item => item.accounting?.resolution !== "unaccountedFor");
  const source = { project, client, spaces, items: reportItems, accountingEvidenceHash };
  const sourceSetHash = v.hash(v.canonical(source));
  const content = { reportKind: "client_summary_physical" as const, source, provenance, sourceSetHash };
  const snapshotHash = v.hash(v.canonical(content));
  return { reportKind: "client_summary_physical", ...source, provenance, sourceSetHash,
    reference: { snapshotID: snapshotHash.slice(0, 32), snapshotHash,
      visibilityScopeID: provenance.visibilityScopeID, profileVersion: "client-summary-physical-v1",
      authorityVersion: provenance.authorityVersion } };
}
export function isClientSummaryPhysicalReportComplete(snapshot: ClientSummaryPhysicalReportSnapshot): boolean {
  return snapshot.client.kind === "known" && snapshot.items.every(item =>
    item.accounting?.resolution === "accountedFor" && "known" in item.category);
}
export function encodeClientSummaryPhysicalReportSnapshot(snapshot: ClientSummaryPhysicalReportSnapshot): string {
  return v.canonical(snapshot);
}
export function encodeClientSummaryPhysicalReportContent(snapshot: ClientSummaryPhysicalReportSnapshot): string {
  const { project, client, spaces, items, accountingEvidenceHash, provenance, sourceSetHash, reportKind } = snapshot;
  return v.canonical({ reportKind, source: { project, client, spaces, items, accountingEvidenceHash }, provenance, sourceSetHash });
}
